zbuf/scanner.go

```diff
@@ -46,10 +45,34 @@ func (s *ScannerStats) Accumulate(ss *ScannerStats) {
 	s.RecordsMatched += ss.RecordsMatched
 }
 
+func ReadersToScanners(ctx context.Context, readers []Reader) ([]Scanner, error) {
+	scanners := make([]Scanner, 0, len(readers))
+	for _, reader := range readers {
+		s, err := NewScanner(ctx, reader, nil, nano.MaxSpan)
+		if err != nil {
+			return nil, err
+		}
+		scanners = append(scanners, s)
+	}
+	return scanners, nil
+}
+
+func ReadersToPullers(ctx context.Context, readers []Reader) ([]Puller, error) {
```

Comment:
Is the main reason to use Scanners to perform this operation, instead of say, zbuf.ReadBatch, so that the improved memory handling of the zngio scanner is used?

Reply:
merge.Proc needs pullers as its parent so this provides the glue when you are directly wiring readers to a merge proc (e.g., file readers that do not make native Scanners like zngio).  This was the minimal change I wanted to make to get this wired up.  We could look into further refactoring.  The other thing I noticed is that stuff that takes a Reader (like zar commands) have to turn the Scanner into a Reader, which forces the copy of Record out of the scanner buffer.  If we changed zar commands to take Pullers instead of Readers then they could spin over the Batch and release it without copying the record out of the Scanner buffer.

Reply:
As suggested by @mattnibs, I'll see about moving the join function into expr to be used separately by zbuf and proc.

Reply:
I ended up moving merger to zbuf instead of expr because it's not an expression and implements zbuf.Puller.